# Files necessary to reproduce the analyses from our manuscript

## Inside the .zip file you will find:
1. Chute_order.txt - a list of all pigs from the fair in the order that they were initially weighed in
2. pig_to_sample.txt - a table that relates the sequenceIDs for the project to the pig they came from, and the day of the fair they were collected
3. swineIAV_dSNPs.txt - the file containing the site and nucleotide expected for lineage I and II viruses from our manuscript
4. ref_cds_loc.txt - a table that is necessary for the iSNV classification script to determine if a iSNV is synonymous or non-synonymous
5. run#_single.txt - the lists of sampleIDs used for the analyses contained in our manuscript
6. swineIAV_avg_cov_table.run#_single.txt - average depth of coverage in samples from singly infected pigs
7. swineIAV_group_count_table.run#_single.txt - the number of dSNPs that were detected in the segments of each sample
8. swineIAV_iSNV_table.run#_single.txt - tables containing the iSNVs detected in our manuscript
9. all_refs.fasta - a fasta file containing the lineage I and II major haplotypes that were used for read mapping in our manuscript
10. S_NS_ratio_summary.txt - the table that contains all S and NS iSNV counts used to calculate the S/NS ratios from our manuscript
